src/tile.rs

```diff
@@ -45,9 +45,59 @@ pub fn calc_tile_id(z: u8, x: u64, y: u64) -> u64 {
     base_id + tile_id
 }
 
+#[must_use]
+pub(crate) fn xyz(tile_id: u64) -> (u8, u64, u64) {
+    if tile_id == 0 {
+        return (0, 0, 0);
+    }
+
+    // Find the zoom level by comparing against pyramid sizes
+    let mut z = 0u8;
+    for (zoom, &pyramid_size) in PYRAMID_SIZE_BY_ZOOM.iter().enumerate() {
+        if tile_id < pyramid_size {
+            z = u8::try_from(zoom - 1).expect("malformed zoom level");
+            break;
+        }
+    }
+
+    // If we didn't find it in the precomputed array, calculate for higher zoom levels
+    if z == 0 && tile_id >= PYRAMID_SIZE_BY_ZOOM[PYRAMID_SIZE_BY_ZOOM.len() - 1] {
+        let last_ind = PYRAMID_SIZE_BY_ZOOM.len() - 1;
+        let mut current_pyramid_size = PYRAMID_SIZE_BY_ZOOM[last_ind];
+        z = u8::try_from(last_ind).expect("malformed zoom level");
+
+        while tile_id >= current_pyramid_size {
+            let tiles_at_zoom = 1_u64 << (z << 1);
+            if tile_id < current_pyramid_size + tiles_at_zoom {
+                break;
+            }
+            current_pyramid_size += tiles_at_zoom;
+            z += 1;
+        }
+    }
+
+    // Calculate base_id for this zoom level
+    let z_ind = usize::from(z);
```

Comment:
this code is the exact duplicate of the `calc_tile_id` - lets not duplicate code, esp math-related and error prone.  Also, I wonder if you can call `calc_tile_id(z, 0, 0)` and use that tileid for something?  Not sure if that makes sense though.